addons/project/models/project.py

```diff
@@ -0,0 +1,832 @@
+
+    @api.multi
+    def copy(self, default=None):
+        if default is None:
+            default = {}
+        context = dict(self.env.context)
+        context['active_test'] = False
+        if not default.get('name'):
+            default.update(name=_("%s (copy)") % (self.name))
+        project = super(Project, self).copy(default)
+        self.with_context(active_test=False).map_tasks(project)
+        return project
+
+    @api.multi
+    def unlink(self):
+        AnalyticAccount = self.env['account.analytic.account']
+        for proj in self:
+            if proj.tasks:
+                raise UserError(_('You cannot delete a project containing tasks. You can either delete all the project\'s tasks and then delete the project or simply deactivate the project.'))
+            if proj.analytic_account_id and not proj.analytic_account_id.line_ids:
+                AnalyticAccount += proj.analytic_account_id
+        res = super(Project, self).unlink()
+        AnalyticAccount.unlink()
+        return res
+
+    def map_tasks(self, project):
+        """ copy and map tasks from old to new project """
+        self.ensure_one()
+        map_task_id = {}
+        for task in self.tasks:
+            # preserve task name and stage, normally altered during copy
+            defaults = {'stage_id': task.stage_id.id,
+                        'name': task.name}
+            map_task_id[task.id] =  task.copy(defaults).id
+        project.write({'tasks':[(6, 0, map_task_id.values())]})
+
+    @api.multi
+    def attachment_tree_view(self):
+        self.ensure_one()
+        tasks = self.env['project.task'].search([('project_id', 'in', self.ids)])
+        domain = [
+             '|',
+             '&', ('res_model', '=', 'project.project'), ('res_id', 'in', self.ids),
+             '&', ('res_model', '=', 'project.task'), ('res_id', 'in', tasks.ids)]
+        return {
+            'name': _('Attachments'),
+            'domain': domain,
+            'res_model': 'ir.attachment',
+            'type': 'ir.actions.act_window',
+            'view_id': False,
+            'view_mode': 'kanban,tree,form',
+            'view_type': 'form',
+            'help': _('''<p class="oe_view_nocontent_create">
+                        Documents are attached to the tasks and issues of your project.</p><p>
+                        Send messages or log internal notes with attachments to link
+                        documents to your project.
+                    </p>'''),
+            'limit': 80,
+            'context': "{'default_res_model': '%s','default_res_id': %d}" % (self._name, self.id)
+        }
+
+
+class Task(models.Model):
+    _name = "project.task"
+    _description = "Task"
+    _date_name = "date_start"
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
+    _order = "priority desc, sequence, date_start, name, id"
+    _mail_post_access = 'read'
+
+    def _get_default_partner(self):
+        context = dict(self.env.context)
```

Comment:
Why copy the context? There's always a context in the new API, and it's not modified.